src/System.Net.Security/src/System/Net/SecureProtocols/_SslState.cs

```diff
@@ -0,0 +1,1745 @@
+             * SSL2_CLIENT_VERSION_MSB (if version greater than SSLv2) ( or v3)
+             * SSL2_CLIENT_VERSION_LSB (if version greater than SSLv2) ( or v3)
+             *
+             * ... SSLv2 SERVER_HELLO ...
+             */
+
+            /* SSLv3 Type 2 Hello starts with
+              * RECORD_LENGTH_MSB  (ignore)
+              * RECORD_LENGTH_LSB  (ignore)
+              * SSL2_CLIENT_HELLO  (must be equal)
+              * SSL2_CLIENT_VERSION_MSB (if version greater than SSLv3)
+              * SSL2_CLIENT_VERSION_LSB (if version greater than SSLv3)
+              *
+              * ... SSLv2 Compatible Hello ...
+              */
+
+            /* SSLv3 Type 3 Hello starts with
+             * 22 (HANDSHAKE MESSAGE)
+             * VERSION MSB
+             * VERSION LSB
+             * RECORD_LENGTH_MSB  (ignore)
+             * RECORD_LENGTH_LSB  (ignore)
+             * HS TYPE (CLIENT_HELLO)
+             * 3 bytes HS record length
+             * HS Version
+             * HS Version
+             */
+
+            /* SSLv2 message codes
+             * SSL_MT_ERROR                0
+             * SSL_MT_CLIENT_HELLO         1
+             * SSL_MT_CLIENT_MASTER_KEY    2
+             * SSL_MT_CLIENT_FINISHED      3
+             * SSL_MT_SERVER_HELLO         4
+             * SSL_MT_SERVER_VERIFY        5
+             * SSL_MT_SERVER_FINISHED      6
+             * SSL_MT_REQUEST_CERTIFICATE  7
+             * SSL_MT_CLIENT_CERTIFICATE   8
+             */
+
+            int version = -1;
+
+            GlobalLog.Assert((bytes != null && bytes.Length > 0), "SslState::DetectFraming()|Header buffer is not allocated will boom shortly.");
+
+            // If the first byte is SSL3 HandShake, then check if we have a SSLv3 Type3 client hello.
+            if (bytes[0] == (byte)FrameType.Handshake || bytes[0] == (byte)FrameType.AppData
+                || bytes[0] == (byte)FrameType.Alert)
+            {
+                if (length < 3)
+                {
+                    return Framing.Invalid;
+                }
+#if TRAVE
+                if (bytes[1] != 3)
+                    GlobalLog.Print("WARNING: SslState::DetectFraming() SSL protocol is > 3, trying SSL3 framing in retail = " + bytes[1].ToString("x", NumberFormatInfo.InvariantInfo));
+#endif
+
+                version = (bytes[1] << 8) | bytes[2];
+                if (version < 0x300 || version >= 0x500)
+                {
+                    return Framing.Invalid;
+                }
+
+                //
+                // This is an SSL3 Framing
+                //
+                return Framing.SinceSSL3;
+            }
+#if TRAVE
+            if ((bytes[0] & 0x80) == 0)
+                // We have a three-byte header format
+                GlobalLog.Print("WARNING: SslState::DetectFraming() SSL v <=2 HELLO has no high bit set for 3 bytes header, we are broken, received byte = " + bytes[0].ToString("x", NumberFormatInfo.InvariantInfo));
+#endif
```

Comment:
I don't understand why we'd expect `bytes[0]` to have the high bit set.  The description in the comment blocks above is that it's the MSB, so that being set requires it to be encoding a length larger than 32767, but the comments keep talking about "3".